docs/design/features/native-hosting.md

```diff
@@ -209,6 +211,7 @@ The function returns specific return code for the first initialized host context
   * In both cases only the properties specified by the new runtime config will be reported on the host context. This is to allow the native host to decide in the "warning" case if it's OK to let the component run or not.
   * In both cases the returned host context can still be used to get a runtime delegate, the properties from the new runtime config will be ignored (as there's no way to modify those in the runtime).
 
+The specified `.runtimeconfig.json` must be for a framework dependent component. That is it must specify at least one shared framework in its `frameworks` section. Self-contained components are not supported.
```

Comment:
All of my questions about this design basically boil down to this: why should `hostfxr_initialize_for_runtime_config` still block on an SCD `.runtimeconfig.json` since the user can now just use `hostfxr_initialize_for_dotnet_command_line` instead?

Reply:
In my mind `hostfxr_initialize_for_runtime_config` is targeted at components (so things like plugins). It's usefulness for apps is questionable due to the fact that it doesn't populate TPA (the default ALC) with application code, only frameworks.

On top of that "self-contained component" is not a thing in .NET Core (yet). There's no SDK support for it. We only allow one runtime in process as well, which complicates this as well.

If we were to allow SCD in `hostfxr_initialize_for_runtime_config` it would have to be either a very limited experience (cumbersome build, only one such component in process, hard to explain, ...) or we would have to deliver lot of new functionality (SDK support, probably at least some basic VS support, open the question of multiple runtimes in process which ignoring some technical difficulties in the runtime would cause lot of problems for diagnostics, ...).

Combined with the fact that the scenario you describe (along with several other users asking for something similar) is somewhere in between app and component (as you correctly pointed out and I didn't realize at the beginning, your scenario is much closure to an app in a way).

I've also seen a few asks on allowing the "get delegate" functionality on pure apps (no component involved), even ones started directly as .NET Core apps (no native host at the beginning).

In theory it's possible to use `hostfxr_initialize_for_dotnet_command_line` on "components", but I view that as "bending the system" - quite a bit in fact. And still, if I would want SCD, then it has to be an app, there's no other supported way to actually build it.

Reply:
In .NET 5, the distinction between classlib and app is going away. There is no netstandard for .NET 5, there's only the `net5.0` TFM. It was always possible to build a component targeting `netcoreappX.Y`. So I don't think that part of the argument is valid.

What I'm trying to point out here is that under this design, `hostfxr_get_runtime_delegate` can now be called from a context that didn't perform this SCD check. Which means all of the existing delegate types (which are all targeted to the component scenario) can now be called against an SCD component. If the artificial block in `hostfxr_initialize_for_runtime_config` is kept in, people are just going to get frustrated and go around it anyway.

Reply:
Using a TFM which targets a specific runtime versus netstandard is not the same thing as treating classlib and app the same way. As you note it was always possible to target classlib for a specific runtime, that's not the limitation. The limitation is that SDK simply doesn't officially support building SCD classlib projects. In a way even `dotnet publish` is sort of weird on a classlib (and has several really weird behaviors). All of this is mostly because nobody went and made sure that SDK has this scenario cleaned up. Introducing support for this in the hosting layer feels wrong if we don't have a good story how to build such a thing.

You're right that we might need to block one of the other supported delegates from this new scenario. But most of them should work. There's a separate question if we should support them there - as COM/WinRT/IJW should never run into the new scenario as those entry points will always go through the `hostfxr_initialize_for_runtime_config`. So maybe it would be better to block those cases (just to avoid having to support weird configurations into the future).

I don't think the block in `hostfxr_initialize_for_runtime_config` is artificial. Removing it is actually quite a bit of new work if we were to support comparable behavior to FDD. Currently this entry point "promises" to not load any application code into the default ALC (to avoid version collisions with potential future loaded components). If we were to maintain that promise we would have to implement the split between app/framework assemblies for SCD (per our other discussion - the `runtimepack` value) and split loading that way. On top of that we would have to support then loading what looks like an SCD component into a secondary load context - currently `AssemblyDependencyResolver` assumes that files next to the component being loaded (and mentioned in its `.deps.json`) are part of the component - but for SCD this would be true for the entire framework. So we would have also implement the support for recognizing framework assemblies in `AssemblyDependencyResolver` (this would be a good thing but still more work). I must also admit I'm a bit hesitant to go this way yet, as it would effectively mean splitting SCD into two halves and there might be small behavior differences which show up. Since this is not a primary scenario for us in .NET 5 (we don't have resources to put into this right now) I want to avoid opening too many new scenarios as there's always a bug tail to be fixed. I'm not against doing this in the future though.

Alternative would be to say that when used on SCD it populates the default ALC with the entire app. But then the usage of it would be quite different. Using the existing entry points which all load into secondary ALCs would be problematic due to potential duplication of assemblies (as mentioned in the doc already). And it would be rather confusing - calling the same function would rather significantly change behavior based on what is the input (not the parameters to the function). This would potentially open weird scenarios in COM/WinRT/IJW as all these entry points call the `hostfxr_initialize_for_runtime_config` and if given SCD-like input would get the new behavior and at the very least get confused.

Reply:
Honest question - what makes a classlib a classlib? Because when I do `dotnet new classlib` with .NET 5 Preview 4 it creates a very simple .csproj that seems to be indistinguishable from an app:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>

</Project>
```
I'm not trying to argue for `hostfxr_initialize_for_runtime_config` to support SCD today, I'm just trying to make sure that everyone realizes all of the additional scenarios that this design is opening up. And if the answer is to block all of these additional scenarios, then this is what I was trying to say when I thought going in this direction was a bad idea. If a significant part of the work in implementing this design is adding more blocks, then I don't think it's a good idea to add this "get delegate" functionality inside of `hostfxr_get_runtime_delegate`.

Reply:
The app project default looks like this:
```XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

</Project>
```

The difference is `<OutputType>Exe</OutputType>`. I do agree that the naming is rather unfortunate (especially in light of Linux/Mac), but it's been like this for many many years so it's not worth changing (at least that's what I've heard about it, I'm not directly involved in this).

This setting drives several behavior differences in the SDK, one of them is that this produces `.runtimeconfig.json`, which you don't by default for classlib (although we now have properties to enable that).

Logically (at least for me) the functionality belongs into `hostfxr_get_runtime_delegate`. In ideal world we would enable most of the cases described above and those we mentioned before. In reality we try to prioritize the important scenarios and explicitly block the rest (the reason for that is basically resourcing - the cost is not just adding it now, but maintaining it for a long time - especially since `hostfxr` is semi-global and thus has a much higher backward compat bar than even the runtime itself in some ways). I don't think that exposing the functionality in a new function helps that much - it might be that the API will be slightly easier to understand (since we can give it scenario specific name and parameters), but it would still have the same limitations. And if we wanted to widen the supported cases in the next release having more APIs tends to make that harder.

Reply:
I tested my host with projects that only have `<EnableDynamicLoading>true</EnableDynamicLoading>` targeting `netcoreapp3.x` with FDD and SCD without any problems.

If this is going to be in `hostfxr_get_runtime_delegate`, then I'm interested in the real reasons behind the blocks. And I expect all the people that run into the blocks will be interested as well.

https://github.com/dotnet/runtime/issues/3660 is already slated in .NET 5.0 and seems to be created explicitly for most of the issues you brought up.

As noted above, SDK support seems to be there already. I never tried to set `EnableDynamicLoading` through VS but if that's missing then that would affect FDD, too.

I thought `hostfxr` is already preventing the loading of multiple runtimes into the process.

Reply:
I don't want my objections to hold this up. Please update the document with the scenarios that should be blocked along with the error message that should be logged so that this can be approved and I can start implementing this.